samples/nrf9160/gnss/src/assistance_minimal.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <zephyr.h>
+#include <sys/timeutil.h>
+#include <settings/settings.h>
+#include <logging/log.h>
+#include <nrf_modem_at.h>
+#include <nrf_modem_gnss.h>
+
+#include "assistance.h"
+#include "factory_almanac.h"
+#include "mcc_location_table.h"
+
+LOG_MODULE_DECLARE(gnss_sample, CONFIG_GNSS_SAMPLE_LOG_LEVEL);
+
+/* (6.1.1980 UTC - 1.1.1970 UTC) */
+#define GPS_TO_UNIX_UTC_OFFSET_SECONDS	(315964800UL)
+/* UTC/GPS time offset as of 1st of January 2017. */
+#define GPS_TO_UTC_LEAP_SECONDS		(18UL)
+#define SEC_PER_MIN			(60UL)
+#define MIN_PER_HOUR			(60UL)
+#define SEC_PER_HOUR			(MIN_PER_HOUR * SEC_PER_MIN)
+#define HOURS_PER_DAY			(24UL)
+#define SEC_PER_DAY			(HOURS_PER_DAY * SEC_PER_HOUR)
+#define DAYS_PER_WEEK			(7UL)
+#define PLMN_STR_MAX_LEN		8 /* MCC + MNC + quotes */
+
+static char almanac_checksum[64];
+
+static int set(const char *key, size_t len_rd, settings_read_cb read_cb, void *cb_arg)
+{
+	int len;
+	int key_len;
+	const char *next;
+
+	if (!key) {
+		return -ENOENT;
+	}
+
+	key_len = settings_name_next(key, &next);
+
+	if (!strncmp(key, "almanac_checksum", key_len)) {
+		len = read_cb(cb_arg, &almanac_checksum, sizeof(almanac_checksum));
+		if (len < sizeof(almanac_checksum)) {
+			LOG_ERR("Failed to read almanac checksum from settings");
+		}
+
+		return 0;
+	}
+
+	return -ENOENT;
+}
+
+static struct settings_handler assistance_settings = {
+	.name = "assistance",
+	.h_set = set,
+};
+
+static void factory_almanac_write(void)
+{
+	int err;
+
+	/* Check if the same almanac has already been written to prevent unnecessary writes
+	 * to flash memory.
+	 */
+	if (!strncmp(almanac_checksum, FACTORY_ALMANAC_CHECKSUM, sizeof(almanac_checksum))) {
+		LOG_INF("Factory almanac has already been written, skipping writing");
+		return;
+	}
+
+	err = nrf_modem_at_printf("AT%%XFILEWRITE=1,\"%s\",\"%s\"",
```

Comment:
According to the factory almanac design specification document "GPS receiver will later overwrite the factory almanac with another that has been either received from the broadcast or provided via assistance." This means that if we have obtained a newer almanac from any source, it will get overwritten here with a stale one. Of course, like you said, that may never happen depending on how the GNSS is used.

Customers are actually advised to update the factory almanac in production, see https://infocenter.nordicsemi.com/index.jsp, thus updating almanacs here should bring no benefit in actual applications, right? That being said this code could still be useful for evaluating the product. And possibly updating it in the application enables delivering a new almanac by FOTA...or can FOTA simply write a new almanac file on the flash directly? I don't know how it works.

Reply:
This is just a sample and the main purpose is to show how time from LTE network can be used to assist GNSS, because that's something that customers have been asking about. But the time alone isn't enough, so a complete solution also needs almanacs and location.

Each user needs to decide what's the suitable way to ensure that also almanacs are available for their application (if they decide to use this form of assistance). But, for the sample to be self-sufficient, it needs to ensure that also almanacs are there.

The almanacs are stored in the file system, so updating that directly using FOTA is most likely not possible. I think AT%XFILEWRITE is the only way for an application to update the almanacs through FOTA.

Reply:
> Each user needs to decide what's the suitable way to ensure that also almanacs are available for their application (if they decide to use this form of assistance). But, for the sample to be self-sufficient, it needs to ensure that also almanacs are there.

Is this clear for our customers? Infocenter is not too verbal about what to put to the device in production and whether to insert something in the app side too like the stuff in this sample.

Reply:
As you suggested in another comment, we should have documentation available about the different assistance methods, their differences and usage. That could cover data use, energy consumption and also implementation aspects, like how to ensure that the device has fresh enough almanac available.

If the application is using periodic mode and GNSS scheduled downloads are disabled, it's likely that GNSS will never get fresh almanacs from satellite broadcast. So, the only way for the device to get new almanacs would be to download it over LTE (it could be done using app FOTA, but it would also be possible to do the download for example using HTTP).

If GNSS scheduled downloads are enabled, GNSS will download almanacs from the broadcast every two weeks, but this consumes a lot of energy (it requires that GNSS receiver is powered on for at least 12,5 minutes, in sub-optimal conditions likely longer).